File: src/main/java/io/github/realyusufismail/armourandtoolsmod/core/init/BlockInit.java
/*
 * Copyright 2022 RealYusufIsmail.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */ 
package io.github.realyusufismail.armourandtoolsmod.core.init;

import io.github.realyusufismail.realyusufismailcore.core.init.GeneralBlock;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.registries.RegistryObject;

import java.util.function.Supplier;

import static io.github.realyusufismail.armourandtoolsmod.ArmourAndToolsMod.MOD_ID;

public class BlockInit {
    public static final DeferredRegister<Block> BLOCKS =
            DeferredRegister.create(ForgeRegistries.BLOCKS, MOD_ID);

    // ores
    public static final RegistryObject<GeneralBlock> RUBY_ORE =
            register("ruby_ore", Blocks.IRON_ORE);
    public static final RegistryObject<GeneralBlock> RAINBOW_ORE =
            register("rainbow_ore", Blocks.DIAMOND_ORE);
    public static final RegistryObject<GeneralBlock> SAPPHIRE_ORE =
            register("sapphire_ore", Blocks.GOLD_ORE);
    public static final RegistryObject<GeneralBlock> GRAPHITE_ORE =
            register("graphite_ore", Blocks.COAL_ORE);
    public static final RegistryObject<GeneralBlock> AQUMARINE_ORE =
            register("aqumarine_ore", Blocks.EMERALD_ORE);

    // deepslate ores
    public static final RegistryObject<GeneralBlock> DEEPSLATE_RUBY_ORE =
            register("deepslate_ruby_ore", Blocks.DEEPSLATE_IRON_ORE);
    public static final RegistryObject<GeneralBlock> DEEPSLATE_RAINBOW_ORE =
            register("deepslate_rainbow_ore", Blocks.DEEPSLATE_DIAMOND_ORE);
    public static final RegistryObject<GeneralBlock> DEEPSLATE_SAPPHIRE_ORE =
            register("deepslate_sapphire_ore", Blocks.DEEPSLATE_GOLD_ORE);
    public static final RegistryObject<GeneralBlock> DEEPSLATE_GRAPHITE_ORE =
            register("deepslate_graphite_ore", Blocks.DEEPSLATE_COAL_ORE);
    public static final RegistryObject<GeneralBlock> DEEPSLATE_AQUMARINE_ORE =
            register("deepslate_aqumarine_ore", Blocks.DEEPSLATE_EMERALD_ORE);

    // blocks
    public static final RegistryObject<GeneralBlock> RUBY_BLOCK =
            register("ruby_block", Blocks.IRON_BLOCK);
    public static final RegistryObject<GeneralBlock> RAINBOW_BLOCK =
            register("rainbow_block", Blocks.DIAMOND_BLOCK);
    public static final RegistryObject<GeneralBlock> SAPPHIRE_BLOCK =
            register("sapphire_block", Blocks.GOLD_BLOCK);
    public static final RegistryObject<GeneralBlock> GRAPHITE_BLOCK =
            register("graphite_block", Blocks.COAL_BLOCK);
    public static final RegistryObject<GeneralBlock> AQUMARINE_BLOCK =
            register("aqumarine_block", Blocks.EMERALD_BLOCK);

    private static RegistryObject<GeneralBlock> register(String name,
            Supplier<GeneralBlock> supplier) {
        RegistryObject<GeneralBlock> blockReg = BLOCKS.register(name, supplier);
        ItemInit.ITEMS.register(name, () -> new BlockItem(blockReg.get(), new Item.Properties()));
        return blockReg;
    }

    private static RegistryObject<GeneralBlock> register(String name, Block existingBlock) {
        return register(name,
                () -> new GeneralBlock(BlockBehaviour.Properties.copy(existingBlock)));
    }
}

File: src/main/kotlin/io/github/realyusufismail/armourandtoolsmod/core/init/BlockInit.kt
/*
 * Copyright 2022 RealYusufIsmail.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */ 
package io.github.realyusufismail.armourandtoolsmod.core.init

import io.github.realyusufismail.armourandtoolsmod.MOD_ID
import io.github.realyusufismail.armourandtoolsmod.core.blocks.CustomArmourCraftingTable
import io.github.realyusufismail.armourandtoolsmod.core.blocks.lit.RainbowLitBlock
import io.github.realyusufismail.armourandtoolsmod.core.blocks.lit.RubyLitBlock
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.state.BlockBehaviour
import net.minecraftforge.registries.DeferredRegister
import net.minecraftforge.registries.ForgeRegistries
import thedarkcolour.kotlinforforge.forge.ObjectHolderDelegate
import thedarkcolour.kotlinforforge.forge.registerObject

object BlockInit {
    val BLOCKS: DeferredRegister<Block> = DeferredRegister.create(ForgeRegistries.BLOCKS, MOD_ID)

    val ORE_BLOCKS: MutableMap<ObjectHolderDelegate<Block>, ObjectHolderDelegate<Item>> =
        mutableMapOf()

    val SMELT_ABLE_BLOCKS: MutableMap<ObjectHolderDelegate<Block>, ObjectHolderDelegate<Item>> =
        mutableMapOf()

    val MINABLE_STONE_PICKAXE_BLOCKS: MutableList<ObjectHolderDelegate<Block>> = mutableListOf()

    val MINABLE_IRON_PICKAXE_BLOCKS: MutableList<ObjectHolderDelegate<Block>> = mutableListOf()

    val MINABLE_GOLD_PICKAXE_BLOCKS: MutableList<ObjectHolderDelegate<Block>> = mutableListOf()

    val MINABLE_DIAMOND_PICKAXE_BLOCKS: MutableList<ObjectHolderDelegate<Block>> = mutableListOf()

    val MINABLE_NETHERITE_PICKAXE_BLOCKS: MutableList<ObjectHolderDelegate<Block>> = mutableListOf()

    // ores
    val RUBY_ORE =
        registerSpecialSmeltAbleBlock(
            "ruby_ore", ItemInit.RUBY, MinableBlockType.IRON_PICKAXE, ::RubyLitBlock)

    val RAINBOW_ORE =
        registerSpecialSmeltAbleBlock(
            "rainbow_ore", ItemInit.RAINBOW, MinableBlockType.DIAMOND_PICKAXE, ::RainbowLitBlock)

    val SAPPHIRE_ORE =
        registerSmeltAbleBlock(
            "sapphire_ore",
            ItemInit.SAPPHIRE,
            BlockBehaviour.Properties.copy(Blocks.GOLD_ORE).requiresCorrectToolForDrops(),
            MinableBlockType.GOLD_PICKAXE)

    val GRAPHITE_ORE =
        registerSmeltAbleBlock(
            "graphite_ore",
            ItemInit.GRAPHITE,
            BlockBehaviour.Properties.copy(Blocks.DIAMOND_ORE).requiresCorrectToolForDrops(),
            MinableBlockType.DIAMOND_PICKAXE)

    val AQUMARINE_ORE =
        registerSmeltAbleBlock(
            "aqumarine_ore",
            ItemInit.AQUMARINE,
            BlockBehaviour.Properties.copy(Blocks.DIAMOND_ORE).requiresCorrectToolForDrops(),
            MinableBlockType.DIAMOND_PICKAXE)

    // deepslate ores
    val DEEPSLATE_RUBY_ORE =
        registerSpecialSmeltAbleBlock(
            "deepslate_ruby_ore", ItemInit.RUBY, MinableBlockType.IRON_PICKAXE, ::RubyLitBlock)

    val DEEPSLATE_RAINBOW_ORE =
        registerSpecialSmeltAbleBlock(
            "deepslate_rainbow_ore",
            ItemInit.RAINBOW,
            MinableBlockType.DIAMOND_PICKAXE,
            ::RainbowLitBlock)

    val DEEPSLATE_SAPPHIRE_ORE =
        registerSmeltAbleBlock(
            "deepslate_sapphire_ore",
            ItemInit.SAPPHIRE,
            BlockBehaviour.Properties.copy(Blocks.DEEPSLATE_GOLD_ORE).requiresCorrectToolForDrops(),
            MinableBlockType.GOLD_PICKAXE)

    val DEEPSLATE_GRAPHITE_ORE =
        registerSmeltAbleBlock(
            "deepslate_graphite_ore",
            ItemInit.GRAPHITE,
            BlockBehaviour.Properties.copy(Blocks.DEEPSLATE_DIAMOND_ORE)
                .requiresCorrectToolForDrops(),
            MinableBlockType.DIAMOND_PICKAXE)

    val DEEPSLATE_AQUMARINE_ORE =
        registerSmeltAbleBlock(
            "deepslate_aqumarine_ore",
            ItemInit.AQUMARINE,
            BlockBehaviour.Properties.copy(Blocks.DEEPSLATE_DIAMOND_ORE)
                .requiresCorrectToolForDrops(),
            MinableBlockType.DIAMOND_PICKAXE)

    // blocks
    val RUBY_BLOCK =
        registerOreBlock(
            "ruby_block",
            ItemInit.RUBY,
            BlockBehaviour.Properties.copy(Blocks.IRON_BLOCK).requiresCorrectToolForDrops(),
            MinableBlockType.IRON_PICKAXE)
    val RAINBOW_BLOCK =
        registerOreBlock(
            "rainbow_block",
            ItemInit.RAINBOW,
            BlockBehaviour.Properties.copy(Blocks.DIAMOND_BLOCK).requiresCorrectToolForDrops(),
            MinableBlockType.DIAMOND_PICKAXE)

    val SAPPHIRE_BLOCK =
        registerOreBlock(
            "sapphire_block",
            ItemInit.SAPPHIRE,
            BlockBehaviour.Properties.copy(Blocks.GOLD_BLOCK).requiresCorrectToolForDrops(),
            MinableBlockType.GOLD_PICKAXE)
    val GRAPHITE_BLOCK =
        registerOreBlock(
            "graphite_block",
            ItemInit.GRAPHITE,
            BlockBehaviour.Properties.copy(Blocks.DIAMOND_BLOCK).requiresCorrectToolForDrops(),
            MinableBlockType.DIAMOND_PICKAXE)

    val AQUMARINE_BLOCK =
        registerOreBlock(
            "aqumarine_block",
            ItemInit.AQUMARINE,
            BlockBehaviour.Properties.copy(Blocks.DIAMOND_BLOCK).requiresCorrectToolForDrops(),
            MinableBlockType.DIAMOND_PICKAXE)

    // custom crafting table
    val CUSTOM_ARMOUR_CRAFTING_TABLE =
        registerSpecial("custom_armour_crafting_table", ::CustomArmourCraftingTable)

    private fun <T : Block> registerSpecial(
        name: String,
        supplier: () -> T,
        minableTool: MinableBlockType = MinableBlockType.STONE_PICKAXE,
    ): ObjectHolderDelegate<T> {
        val blockReg = BLOCKS.registerObject(name, supplier)
        ItemInit.ITEMS.registerObject(name) { BlockItem(blockReg.get(), Item.Properties()) }
        checkTypeOfMinableBlock(blockReg as ObjectHolderDelegate<Block>, minableTool)
        return blockReg
    }

    private fun registerSpecialSmeltAbleBlock(
        name: String,
        associatedOreIngot: ObjectHolderDelegate<Item>,
        minableTool: MinableBlockType,
        supplier: () -> Block
    ): ObjectHolderDelegate<Block> {

        val blockReg = BLOCKS.registerObject(name, supplier)
        ItemInit.ITEMS.registerObject(name) { BlockItem(blockReg.get(), Item.Properties()) }

        SMELT_ABLE_BLOCKS[blockReg] = associatedOreIngot
        checkTypeOfMinableBlock(blockReg, minableTool)

        return blockReg
    }

    private fun registerOreBlock(
        name: String,
        associatedOreIngot: ObjectHolderDelegate<Item>,
        property: BlockBehaviour.Properties,
        minableTool: MinableBlockType
    ): ObjectHolderDelegate<Block> {

        val blockReg = BLOCKS.registerObject(name) { Block(property) }

        ItemInit.ITEMS.registerObject(name) { BlockItem(blockReg.get(), Item.Properties()) }

        ORE_BLOCKS[blockReg] = associatedOreIngot
        checkTypeOfMinableBlock(blockReg, minableTool)

        return blockReg
    }

    private fun registerSmeltAbleBlock(
        name: String,
        associatedOreIngot: ObjectHolderDelegate<Item>,
        property: BlockBehaviour.Properties,
        minableTool: MinableBlockType
    ): ObjectHolderDelegate<Block> {

        val blockReg = BLOCKS.registerObject(name) { Block(property) }
        ItemInit.ITEMS.registerObject(name) { BlockItem(blockReg.get(), Item.Properties()) }

        SMELT_ABLE_BLOCKS[blockReg] = associatedOreIngot
        checkTypeOfMinableBlock(blockReg, minableTool)

        return blockReg
    }

    private fun checkTypeOfMinableBlock(
        block: ObjectHolderDelegate<Block>,
        type: MinableBlockType
    ) {
        when (type) {
            MinableBlockType.STONE_PICKAXE -> MINABLE_STONE_PICKAXE_BLOCKS.add(block)
            MinableBlockType.IRON_PICKAXE -> MINABLE_IRON_PICKAXE_BLOCKS.add(block)
            MinableBlockType.GOLD_PICKAXE -> MINABLE_GOLD_PICKAXE_BLOCKS.add(block)
            MinableBlockType.DIAMOND_PICKAXE -> MINABLE_DIAMOND_PICKAXE_BLOCKS.add(block)
            MinableBlockType.NETHERITE_PICKAXE -> MINABLE_NETHERITE_PICKAXE_BLOCKS.add(block)
        }
    }
}
